src/std/misc/rwlock.ss

```diff
@@ -0,0 +1,101 @@
+;;; -*- Gerbil -*-
+;;; © vyzo
+;;; read-write locks
+(import :gerbil/gambit/threads
+        :std/sugar)
+(export make-rwlock
+        rwlock?
+        rwlock-read-lock!
+        rwlock-read-unlock!
+        rwlock-write-lock!
+        rwlock-write-unlock!
+        with-read-lock
+        with-write-lock)
+(declare (not safe))
+
+(defstruct rwlock (mx cv readers writer writers-waiting)
+  final: #t unchecked: #t
+  constructor: :init!)
+
+(defmethod {:init! rwlock}
+  (lambda (self (name 'rwlock))
+    (let ((mx (make-mutex name))
+          (cv (make-condition-variable name)))
+      (set! (&rwlock-mx self) mx)
+      (set! (&rwlock-cv self) cv)
+      (set! (&rwlock-readers self) 0)
+      (set! (&rwlock-writer self) #f)
+      (set! (&rwlock-writers-waiting self) 0))))
+
+(def (rwlock-read-lock! rw)
+  (with ((rwlock mx cv) rw)
+    (mutex-lock! mx)
+    (let lp ()
+      (if (or (&rwlock-writer rw) (fx> (&rwlock-writers-waiting rw) 0))
+        (begin
+          (mutex-unlock! mx cv)
+          (lp))
```

Comment:
Wait, you have a mutex-unlock! in the look, but the lock! outside the loop? How are things balanced?

Also—is that a busy polling, or is mutex-lock! clever about it? I assume the latter.

Reply:
And, shouldn't you be queueing the readers? Or is the queue implicit in the mutex-lock enough?

Reply:
it is the second form of unlock, passing the cv.
Thats gambit's way of waiting on cvs, it atomically unlocks, waits, and then relocks before returning.

Reply:
the readet queue is implicit on the lock, no reason to make it explicit.